pkg/apis/pipeline/v1/result_validation.go

```diff
@@ -111,6 +111,53 @@ func (tr TaskResult) validateValue(ctx context.Context) (errs *apis.FieldError)
 	return errs
 }
 
+// Validate implements apis.Validatable
+func (sr StepResult) Validate(ctx context.Context) (errs *apis.FieldError) {
+	if !resultNameFormatRegex.MatchString(sr.Name) {
+		return apis.ErrInvalidKeyName(sr.Name, "name", fmt.Sprintf("Name must consist of alphanumeric characters, '-', '_', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my-name',  or 'my_name', regex used for validation is '%s')", ResultNameFormat))
+	}
+
+	switch {
+	case sr.Type == ResultsTypeObject:
+		errs := validateObjectStepResult(sr)
+		return errs
+	case sr.Type == ResultsTypeArray:
+		return nil
+	// Resources created before the result. Type was introduced may not have Type set
+	// and should be considered valid
+	case sr.Type == "":
+		return nil
+	// By default, the result type is string
+	case sr.Type != ResultsTypeString:
+		return apis.ErrInvalidValue(sr.Type, "type", "type must be string")
+	}
```

Comment:
Error message here isn't accurate. You probably want something like:

```suggestion
	case sr.Type == ResultsTypeString:
		return nil
	default:
		return apis.ErrInvalidValue(sr.Type, "type", fmt.Sprintf("invalid type %s", sr.Type))
	}
```